pkg/mcs/resource_manager/server/server.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"flag"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"sync"
+	"sync/atomic"
+	"syscall"
+
+	grpcprometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/soheilhy/cmux"
+	"github.com/spf13/cobra"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/mcs/registry"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"github.com/tikv/pd/pkg/utils/logutil"
+	"github.com/tikv/pd/pkg/utils/metricutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+// Server is the resource manager server, and it implements bs.Server.
+// nolint
+type Server struct {
+	cfg *Config
+	mux cmux.CMux
+	// Server state. 0 is not serving, 1 is serving.
+	isServing int64
+	ctx       context.Context
+	name      string
+
+	etcdClient *clientv3.Client
+	httpClient *http.Client
+
+	grpcServer *grpc.Server
+	httpServer *http.Server
+
+	serverLoopCtx    context.Context
+	serverLoopCancel func()
+	serverLoopWg     sync.WaitGroup
+	// Callback functions for different stages
+	// startCallbacks will be called after the server is started.
+	startCallbacks []func()
+	// primaryCallbacks will be called after the server becomes leader.
+	primaryCallbacks []func(context.Context)
+	registry         *registry.ServiceRegistry
+}
+
+// Name returns the unique etcd Name for this server in etcd cluster.
+func (s *Server) Name() string {
+	return s.name
+}
+
+// Context returns the context.
+func (s *Server) Context() context.Context {
+	return s.ctx
+}
+
+// Run runs the pd server.
+func (s *Server) Run() error {
+	if err := s.startClient(); err != nil {
+		return err
+	}
+	if err := s.startServer(s.ctx); err != nil {
+		return err
+	}
+	return s.mux.Serve()
+}
+
+// Close closes the server.
+func (s *Server) Close() {
+	if !atomic.CompareAndSwapInt64(&s.isServing, 1, 0) {
+		// server is already closed
+		return
+	}
+
+	log.Info("closing server")
+
+	s.serverLoopCancel()
```

Comment:
How to close the server?

Reply:
add functions for closing grpc server and http server